src/blcu/download.go

```diff
@@ -17,34 +17,34 @@ type downloadRequest struct {
 	Board string `json:"board"`
 }
 
-type downloadData struct {
-	Board   string
-	Payload []byte
-}
-
-func (blcu *BLCU) handleDownload(payload json.RawMessage) (downloadData, error) {
+func (blcu *BLCU) download(payload json.RawMessage) (string, DownloadResult) {
 	blcu.trace.Debug().Msg("Handling download")
 	var request downloadRequest
 	if err := json.Unmarshal(payload, &request); err != nil {
 		blcu.trace.Error().Err(err).Stack().Msg("Unmarshal payload")
-		return downloadData{}, err
+		return "", DownloadResult{Err: err}
 	}
 
 	if err := blcu.requestDownload(request.Board); err != nil {
 		blcu.trace.Error().Err(err).Stack().Msg("Request download")
-		return downloadData{}, err
+		return request.Board, DownloadResult{Err: err}
 	}
 
-	buffer := bytes.NewBuffer([]byte{})
-	if err := blcu.ReadTFTP(buffer); err != nil {
-		blcu.trace.Error().Err(err).Stack().Msg("Read TFTP")
-		return downloadData{}, err
+	download, err := blcu.ReadTFTP()
```

Comment:
I would do the following: ReadTFTP blocks until it downloads or errors and it accepts a callback (onProgress), where you receive the progress and can whatever you want to do with it.